### PR TITLE
Implement gel.Record type

### PR DESCRIPTION
**Pull request by 1st1**
It's returned from `client.query_sql()` and represents SQL rows.

Currently it supports:

* integer indexing - fetch columns by position
* string indexing - fetch columns by name
* as_dict() method - render record as a dict

ToDo:

* add __iter__

### CODE REVIEW

File: gel/datatypes/object.c [hunk 1]
```diff
@@ -284,7 +265,7 @@ object_repr(EdgeObject *o)
 
     if (_EdgeGeneric_RenderItems(&writer,
                                  (PyObject *)o, o->desc,
-                                 o->ob_item, Py_SIZE(o), 1, 0) < 0)
+                                 o->ob_item, Py_SIZE(o), 1, 1, 0) < 0)
```
Comment (msullivan) [inline gel/datatypes/object.c:268]: You don't really need to fix this now, but this sort of list of boolean arguments ought to either have comments saying what each of the argument names are or should use #defines/enums or flag bits or something. `1, 1, 0` is inscrutable.

This is an existing problem, though, so it's fine to skip fixing it now.

File: gel/datatypes/record.c [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+
+static PyObject *
+record_repr(EdgeRecord *o)
+{
+    _PyUnicodeWriter writer;
+    _PyUnicodeWriter_Init(&writer);
+    writer.overallocate = 1;
+
+    if (_PyUnicodeWriter_WriteASCIIString(&writer, "Row{", 4) < 0) {
```
Comment (msullivan) [inline gel/datatypes/record.c:203]: Should this be `Record{`?

File: tests/test_async_query.py [hunk 3]
```diff
@@ -1151,3 +1151,13 @@ async def test_transaction_state(self):
...
+
+    # async def test_async_query_sql_01(self):
+    #     res = await self.client.query_sql("SELECT 1")
+    #     self.assertEqual(res[0].as_dict(), {'col~1': 1})
+
+    #     res = await self.client.query_sql("SELECT 1 as aa")
+    #     self.assertEqual(res[0].as_dict(), {'aa': 1})
+
+    #     res = await self.client.query_sql("SELECT FROM generate_series(0, 1)")
+    #     self.assertEqual(res[0].as_dict(), {})
```
Comment (msullivan) [inline tests/test_async_query.py:1163]: Use `xfail` or `skip` instead of commenting out

**Review by msullivan — APPROVED** [review]
Looks reasonable. Kind of a lot of duplication, but you said that merging it would be worse?

**Comment by 1st1** [comment]
@msullivan 

> Kind of a lot of duplication

What's duplicated? Besides the Record codec looking like a stripped down Object codec?

File: tests/test_async_query.py (hunk 3 above)
Comment (1st1) [reply tests/test_async_query.py]: sure

File: gel/datatypes/record.c (hunk 2 above)
Comment (1st1) [reply gel/datatypes/record.c]: fixed

File: gel/datatypes/object.c (hunk 1 above)
Comment (1st1) [reply gel/datatypes/object.c]: Fixed in [Apply better style for passing flags](https://github.com/edgedb/edgedb-python/pull/560/commits/c62afea58c1bcb46759038e2354d1437ea20dd8b)

Comment (msullivan) [reply gel/datatypes/object.c]: Much nicer

**Pull request merged**